File: frontend/scholar-1.0.0/education-ui/src/app/core/services/explore.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import type {
  ExploreCollectionCardDto,
  ExploreCollectionDetailDto,
  ExploreOptionDto,
  ExplorePortfolioDetailDto,
  ExplorePortfolioCardDto,
  ExploreProjectDetailDto,
  ExploreProjectCardDto,
  ExploreSort,
  ExploreVisibilityFilter,
  SkillSummaryDto,
} from '../models/api.models';

export type ExploreFilters = {
  q?: string;
  jobTitle?: string;
  family?: string;
  category?: string;
  skillIds?: number[];
  visibility?: ExploreVisibilityFilter;
  sort?: ExploreSort;
};

@Injectable({ providedIn: 'root' })
export class ExploreService {
  private readonly http = inject(HttpClient);

  private api(path: string): string {
    const base = (environment.apiUrl ?? '').trim().replace(/\/$/, '');
    const p = path.startsWith('/') ? path : `/${path}`;
    return base ? `${base}${p}` : p;
  }

  getFamilies(): Observable<ExploreOptionDto[]> {
    return this.http.get<ExploreOptionDto[]>(this.api('/api/explore/families'));
  }

  getSkillCategories(): Observable<ExploreOptionDto[]> {
    return this.http.get<ExploreOptionDto[]>(this.api('/api/explore/skill-categories'));
  }

  searchSkills(query?: string): Observable<SkillSummaryDto[]> {
    let params = new HttpParams();
    const q = query?.trim();
    if (q) {
      params = params.set('q', q);
    }
    return this.http.get<SkillSummaryDto[]>(this.api('/api/explore/skills'), { params });
  }

  searchPortfolios(query: ExploreFilters): Observable<ExplorePortfolioCardDto[]> {
    return this.http.get<ExplorePortfolioCardDto[]>(this.api('/api/explore/portfolios'), {
      params: this.buildParams(query),
    });
  }

  searchProjects(query: ExploreFilters): Observable<ExploreProjectCardDto[]> {
    return this.http.get<ExploreProjectCardDto[]>(this.api('/api/explore/projects'), {
      params: this.buildParams(query),
    });
  }

  searchCollections(query: ExploreFilters): Observable<ExploreCollectionCardDto[]> {
    return this.http.get<ExploreCollectionCardDto[]>(this.api('/api/explore/collections'), {
      params: this.buildParams(query),
    });
  }

  getPortfolioDetail(portfolioId: number): Observable<ExplorePortfolioDetailDto> {
    return this.http.get<ExplorePortfolioDetailDto>(this.api(`/api/explore/portfolios/${portfolioId}`));
  }

  getProjectDetail(projectId: number): Observable<ExploreProjectDetailDto> {
    return this.http.get<ExploreProjectDetailDto>(this.api(`/api/explore/projects/${projectId}`));
  }

  getCollectionDetail(collectionId: number): Observable<ExploreCollectionDetailDto> {
    return this.http.get<ExploreCollectionDetailDto>(this.api(`/api/explore/collections/${collectionId}`));
  }

  getPortfolioCollections(portfolioId: number): Observable<ExploreCollectionCardDto[]> {
    return this.http.get<ExploreCollectionCardDto[]>(this.api(`/api/explore/portfolios/${portfolioId}/collections`));
  }

  private buildParams(query: ExploreFilters): HttpParams {
    let params = new HttpParams();
    if (query.family?.trim()) {
      params = params.set('family', query.family.trim());
    }
    if (query.category?.trim()) {
      params = params.set('category', query.category.trim());
    }
    if (query.q?.trim()) {
      params = params.set('q', query.q.trim());
    }
    if (query.jobTitle?.trim()) {
      params = params.set('jobTitle', query.jobTitle.trim());
    }
    if (query.visibility?.trim()) {
      params = params.set('visibility', query.visibility.trim());
    }
    if (query.sort?.trim()) {
      params = params.set('sort', query.sort.trim());
    }
    for (const skillId of query.skillIds ?? []) {
      params = params.append('skillIds', String(skillId));
    }
    return params;
  }
}

File: spring/backend/src/main/java/com/education/platform/services/interfaces/explore/ExploreService.java
package com.education.platform.services.interfaces.explore;

import com.education.platform.dto.explore.ExploreOptionDto;
import com.education.platform.dto.explore.ExploreCollectionDetailResponse;
import com.education.platform.dto.explore.ExploreCollectionResponse;
import com.education.platform.dto.explore.ExplorePortfolioDetailResponse;
import com.education.platform.dto.explore.ExplorePortfolioCardResponse;
import com.education.platform.dto.explore.ExploreProjectDetailResponse;
import com.education.platform.dto.explore.ExploreProjectCardResponse;
import com.education.platform.dto.explore.ExploreSort;
import com.education.platform.dto.explore.ExploreVisibilityFilter;
import com.education.platform.dto.portfolio.SkillSummaryDto;
import com.education.platform.dto.portfolio.ai.DeveloperFamily;
import com.education.platform.entities.User;
import com.education.platform.entities.portfolio.SkillCategory;

import java.util.List;

public interface ExploreService {

    List<ExploreOptionDto> listFamilies();

    List<ExploreOptionDto> listSkillCategories();

    List<SkillSummaryDto> searchSkills(User currentUser, String query);

    List<ExplorePortfolioCardResponse> searchPortfolios(
            User currentUser,
            DeveloperFamily family,
            SkillCategory category,
            List<Long> skillIds,
            String query,
            String jobTitle,
            ExploreVisibilityFilter visibility,
            ExploreSort sort);

    List<ExploreProjectCardResponse> searchProjects(
            User currentUser,
            DeveloperFamily family,
            SkillCategory category,
            List<Long> skillIds,
            String query,
            ExploreVisibilityFilter visibility,
            ExploreSort sort);

    List<ExploreCollectionResponse> searchCollections(
            User currentUser,
            DeveloperFamily family,
            SkillCategory category,
            List<Long> skillIds,
            String query,
            ExploreVisibilityFilter visibility,
            ExploreSort sort);

    ExplorePortfolioDetailResponse getPortfolioDetail(User currentUser, Long portfolioId);

    ExploreProjectDetailResponse getProjectDetail(User currentUser, Long projectId);

    ExploreCollectionDetailResponse getCollectionDetail(User currentUser, Long collectionId);

    List<ExploreCollectionResponse> getPortfolioCollections(User currentUser, Long portfolioId);
}
